bonsai-core/src/main/scala/com/stripe/bonsai/Layout.scala

```diff
@@ -4,31 +4,45 @@ import scala.language.implicitConversions
 
 import scala.reflect.ClassTag
 
+import java.io.{ DataOutput, DataInput }
+
 import com.stripe.bonsai.layout._
 
 trait Layout[A] {
   def newBuilder: VecBuilder[A]
 
+  def write(vec: Vec[A], out: DataOutput): Unit
+
+  def read(in: DataInput): Vec[A]
+
+  def isSafeToCast(vec: Vec[_]): Boolean
+
```

Comment:
Part of the problem is that the `recast`s below actually use a very specific return type which is required so that we *know* we get the right type back. We could add it here, but I think the duplication would live on :( I'm not super satisified with this in general though... I'm not sure how much we need to optimize writes anyways.